crates/bevy_ecs/src/query/state.rs

```diff
@@ -1391,37 +1377,62 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
 }
 
 /// An error that occurs when retrieving a specific [`Entity`]'s query result from [`Query`](crate::system::Query) or [`QueryState`].
-#[derive(Debug, PartialEq, Eq, Clone)]
-pub enum QueryEntityError {
-    /// The given [`Entity`]'s components do not match the query, see [`QueryEntityMismatchDetail`].
-    QueryDoesNotMatch(QueryEntityMismatchDetail, QueryEntityErrorDetail),
-    /// The given [`Entity`] does not exist.
-    NoSuchEntity(QueryEntityErrorDetail),
-    /// The [`Entity`] was requested mutably more than once.
-    ///
-    /// See [`QueryState::get_many_mut`] for an example.
-    AliasedMutability(QueryEntityErrorDetail),
+#[derive(Debug, PartialEq, Eq, Clone, Error)]
+#[error("Error querying for {entity:?} in query {error_type_info}: {kind}")]
+pub struct QueryEntityError<ET> {
+    /// Specific requested entity when encountering the error.
+    pub entity: Entity,
+    /// The query's type.
+    error_type_info: ET,
+    /// The kind of error preventing access to the entity.
+    pub kind: ErrorKind,
 }
-
-/// Additional information in the context of a [`QueryEntityError`].
-#[derive(Debug, PartialEq, Eq, Clone)]
-pub struct QueryEntityErrorDetail {
-    /// Specific entity which was requested when encountering the error.
-    pub requested_entity: Entity,
-    /// Representation of the query's type.
-    pub query_type: &'static str,
+impl QueryEntityError<()> {
+    fn new<Q: WorldQuery, F: ReadOnlyWorldQuery>(
+        entity: Entity,
+        error_type: ErrorKind,
+    ) -> QueryEntityError<ErrorTypeInfo<Q, F>> {
+        QueryEntityError {
+            entity,
+            kind: error_type,
+            error_type_info: ErrorTypeInfo::new(),
+        }
+    }
+}
+type ErrInfo<Q, F> = ErrorTypeInfo<Q, F>;
+#[derive(PartialEq, Eq, Clone, Copy)]
+pub struct ErrorTypeInfo<Q: WorldQuery, F: ReadOnlyWorldQuery>(PhantomData<fn(Q, F)>);
+impl<Q: WorldQuery, F: ReadOnlyWorldQuery> ErrorTypeInfo<Q, F> {
+    fn new() -> Self {
+        ErrorTypeInfo(PhantomData)
+    }
+}
+impl<Q: WorldQuery, F: ReadOnlyWorldQuery> fmt::Debug for ErrorTypeInfo<Q, F> {
```

Comment:
```suggestion
impl<Q: WorldQuery, F: ReadOnlyWorldQuery> fmt::Display for ErrorTypeInfo<Q, F> {
```
I'm not exactly sure how this compiles. at line 1381
```rust
#[error("Error querying for {entity:?} in query {error_type_info}: {kind}")]
```
It calls the display implementation of `ErrorTypeInfo`, yet there is no such impl.